guides/source/security.md

```diff
@@ -80,31 +80,35 @@ The main objective of most attackers is to make money. The underground prices fo
 
-* _Critical data should not be stored in session_. If the user clears their cookies or closes the browser, they will be lost. And with a client-side session storage, the user can read the data.
+* _Sensitive data should be stored in a session_. Client can read the local storages, and copy them at will.
+
+* _Large objects should not be stored in a session_. Instead store your objects in the database and at most, save their id in the session. This will spare you from unnecessary synchronization headaches and also, requires less session storage space.
 
 ### Session Storage
 
-NOTE: _Rails provides several storage mechanisms for the session hashes. The most important is `ActionDispatch::Session::CookieStore`._
+NOTE: _Rails provides several storage mechanisms for the session hashes. The most important one is `ActionDispatch::Session::CookieStore`._
+
+Rails `ActionDispatch::Session::CookieStore` saves the session hash in a cookie on the client-side. The server retrieves the session hash from the cookie, which effectively acts as the session ID. When you use cookies as the session storage, it is good to understand how they work.
+
+* Cookies imply a strict size limit of 4kB. Use the cookie only for data which is relevant for the session.
+
+* The client has full read and write access to the cookie. Client can copy cookies from computer to another. _In general, do not store anything sensitive or security related in the session._
+
+* The server can set expiration time for the cookie, but the client may delete the cookie and its contents before that. Also, the client may preserve cookie contents even for expired cookies. Persist the data you want to persist on the server side.
 
-Rails 2 introduced a new default session storage, CookieStore. CookieStore saves the session hash directly in a cookie on the client-side. The server retrieves the session hash from the cookie and eliminates the need for a session ID. That will greatly increase the speed of the application, but it is a controversial storage option and you have to think about the security implications of it:
+#### Security measures to protect your cookies
 
-* Cookies imply a strict size limit of 4kB. This is fine as you should not store large amounts of data in a session anyway, as described before. _Storing the current user's database id in a session is usually ok_.
+Setup secrets in config/secret.yml for all your environments.
 
-* The client can see everything you store in a session, because it is stored in clear-text (actually Base64-encoded, so not encrypted). So, of course, _you don't want to store any secrets here_. To prevent session hash tampering, a digest is calculated from the session with a server-side secret (`secrets.secret_token`) and inserted into the end of the cookie.
+When choosing secrets, make sure they are long and random enough, and consequently, harder to break. _Do not use a trivial secrets, i.e. a dictionary word, or secrets shorter than 30 characters_.
```

Comment:
Maybe simpler to collapse this and below down to 'secrets must be long and random. use `rails secret`.'